### PR TITLE
Jail.RemoveCells before logout and switching account

### PR DESCRIPTION
**Pull request by oleg-raev**
Added methods:
- Jail.RemoveCells
- JailCell.Stop
- Loop.Stop
 
Added call of method RemoveCells at begin of Logout action and begin of SelectAccount action of geth API

Added tests:
- (api_test.go) TestJailRemoveCells
- (api_test.go) TestLogoutRemovesCells
- (jail_test.go) TestJailRemoveCells
- (jail_cell_test.go) TestJailCellStopSetIntervalBeforeStop
- (jail_cell_test.go) TestJailCellStopSetIntervalAfterStop
- (jail_cell_test.go) TestJailCellStopTwice

Issue #260 - setInterval keeps on posting updates after the account has been switched

### CODE REVIEW

**Comment by divan** [comment]
@oleg-raev thanks for terrific work. This code is not easy to grasp quickly.
However, this functionality already has been implemented in https://github.com/status-im/status-go/pull/376/commits, and we're going to merge it today. The approach there is similar to yours, but uses `context` package for cancelling cells and makes code a bit easier. What I really like is a test/code ratio, and I'd really revisit it after merging (we going to merge a bunch of pending refactoring branches).

For now, marking this as blocked by https://github.com/status-im/status-go/pull/376

**Comment by oleg-raev** [comment]
@divan Thank you for your update and quick merging your PR.

> The approach there is similar to yours, but uses context package for cancelling cells
Yes, it looks better and easier.

I should merge conflicts and update the pull request, right?
What about failing check? Is it a big problem?

**Comment by divan** [comment]
> I should resolve conflicts and update the pull request, right?
Umm, not yet, the changes were huge (including file renames). We're about to merge another PR, related to tests. Once it's done, we'll revisit this PR and probably will manually cherry pick some changes (like logout API, and especially tests).

**Comment by tiabc** [comment]
Thank you very much for the contribution, @oleg-raev!

We may either cherry-pick needed commits from your branch manually, or you can groom this branch yourself after #375 is merged if you have time. Which one is preferred for you?

Added a zenhub dependency on #371 as per @divan's comment.

**Comment by oleg-raev** [comment]
>We may either cherry-pick needed commits from your branch manually

There is only one commit.

>you can groom this branch yourself after #375 is merged if you have time

Yes. Let's go this way.

**Comment by tiabc** [comment]
@oleg-raev please, proceed.

**Comment by oleg-raev** [comment]
@tiabc conflicts resolved

File: e2e/api/api_test.go [hunk 1]
```diff
@@ -118,3 +135,85 @@ func (s *APITestSuite) TestRaceConditions() {
 	time.Sleep(2 * time.Second) // so that we see some logs
 	s.api.StopNode()            // just in case we have a node running
 }
+
+func (s *APITestSuite) TestJailStop() {
```
Comment (tiabc) [inline e2e/api/api_test.go]: `TestCellsRemovedAfterSwitchAccount`

File: e2e/api/api_test.go [hunk 2]
```diff
@@ -118,3 +135,85 @@ func (s *APITestSuite) TestRaceConditions() {
...
+
+	address2, _, _, err := s.api.AccountManager().CreateAccount(TestConfig.Account2.Password)
+	require.NoError(err)
+
+	err = s.api.SelectAccount(address1, TestConfig.Account1.Password)
+	require.NoError(err)
+
+	for i := 0; i < itersCount; i++ {
+		wg.Add(1)
+		go func(id int) {
```
Comment (tiabc) [inline e2e/api/api_test.go]: Why doing it in a goroutine?

File: e2e/api/api_test.go [hunk 3]
```diff
@@ -118,3 +135,85 @@ func (s *APITestSuite) TestRaceConditions() {
...
+	address1, _, _, err := s.api.AccountManager().CreateAccount(TestConfig.Account1.Password)
+	require.NoError(err)
+
+	err = s.api.SelectAccount(address1, TestConfig.Account1.Password)
+	require.NoError(err)
+
+	s.api.JailManager().Parse(testChatID, ``)
+
+	_, err = s.api.JailManager().NewCell(testChatID)
+	require.NoError(err)
```
Comment (tiabc) [inline e2e/api/api_test.go]: Sorry, what's the meaning of this assertion?

File: e2e/jail/jail_test.go [hunk 4]
```diff
@@ -157,3 +159,49 @@ func (s *JailTestSuite) TestEventSignal() {
 	expectedResponse := `{"jsonrpc":"2.0","result":true}`
 	s.Equal(expectedResponse, response)
 }
+
+func (s *JailTestSuite) TestJailStop() {
+	const loopLen = 5
+	var wg sync.WaitGroup
```
Comment (tiabc) [inline e2e/jail/jail_test.go]: Again: why goroutines?

File: e2e/jail/jail_test.go [hunk 5]
```diff
@@ -157,3 +159,49 @@ func (s *JailTestSuite) TestEventSignal() {
...
+				setInterval(function(){
+					counter++;
+				}, 1000);
+			`)
+		}(i)
+	}
+
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
```
Comment (tiabc) [inline e2e/jail/jail_test.go]: Won't it be easier to read without `defer`?

File: e2e/jail/jail_test.go [hunk 6]
```diff
@@ -157,3 +159,49 @@ func (s *JailTestSuite) TestEventSignal() {
 	expectedResponse := `{"jsonrpc":"2.0","result":true}`
 	s.Equal(expectedResponse, response)
 }
+
+func (s *JailTestSuite) TestJailStop() {
```
Comment (tiabc) [inline e2e/jail/jail_test.go]: `TestJailCellsRemovedAfterStop`

File: geth/api/api.go [hunk 7]
```diff
@@ -146,11 +146,13 @@ func (api *StatusAPI) VerifyAccountPassword(keyStoreDir, address, password strin
 // using provided password. Once verification is done, decrypted key is injected into Whisper (as a single identity,
 // all previous identities are removed).
 func (api *StatusAPI) SelectAccount(address, password string) error {
+	api.b.jailManager.Stop()
```
Comment (tiabc) [inline geth/api/api.go]: Hm. Currently `JailManager.Stop` doesn't stop Jail, it just resets it to an initial state. Can you make a comment above the original method like `// FIXME(oleg-raev): This method doesn't make stop, it rather resets its cells to an initial state and should be properly renamed, for example: ResetCells`?

File: geth/jail/cell_test.go [hunk 8]
```diff
@@ -19,6 +20,10 @@ var (
 	baseStatusJSCode = string(static.MustAsset("testdata/jail/status.js"))
 )
 
+func TestCellTestSuite(t *testing.T) {
+	suite.Run(t, new(CellTestSuite))
+}
```
Comment (tiabc) [inline geth/jail/cell_test.go]: +1

File: e2e/api/api_test.go (hunk 3 above)
Comment (oleg-raev) [reply e2e/api/api_test.go]: Exactly, you are right. Cell created by previous call `s.api.JailManager().Parse(...`

**Comment by oleg-raev** [comment]
@tiabc fixed

File: geth/api/api.go [hunk 9]
```diff
@@ -146,11 +146,15 @@ func (api *StatusAPI) VerifyAccountPassword(keyStoreDir, address, password strin
 // using provided password. Once verification is done, decrypted key is injected into Whisper (as a single identity,
 // all previous identities are removed).
 func (api *StatusAPI) SelectAccount(address, password string) error {
+	// FIXME(oleg-raev): This method doesn't make stop, it rather resets its cells to an initial state
+	// and should be properly renamed, for example: ResetCells
```
Comment (tiabc) [inline geth/api/api.go]: I meant to the method declaration. But never mind, it's fine.

**Comment by tiabc** [comment]
Merging without another review as this PR has been hanging too long.
There's a failure in CI caused by a timeout: https://travis-ci.org/status-im/status-go/builds/288617791#L527
These tests pass locally and I'll file a PR for this particular timeout soon.

**Comment by tiabc** [comment]
@oleg-raev thanks one more time for the successful contribution! We'll have a conversation about you joining core contributors and I'll get back to you with the result.

**Pull request merged**